File: 1601-1700/1651-1660/1657M_Determine_if_Two_Strings_Are_Close/map.js
// 1657. Determine if Two Strings Are Close
// https://leetcode.com/problems/determine-if-two-strings-are-close/description/
/**
 * @param {string} word1
 * @param {string} word2
 * @return {boolean}
 */
var closeStrings = function (word1, word2) {
  if (word1.length !== word2.length) return false;
  const map1 = new Map();
  const map2 = new Map();
  for (let i = 0; i < word1.length; i++) {
    let count = map1.get(word1[i]) ?? 0;
    map1.set(word1[i], count + 1);
    count = map2.get(word2[i]) ?? 0;
    map2.set(word2[i], count + 1);
  }
  const letter1 = [...map1.keys()].sort().join('');
  const letter2 = [...map2.keys()].sort().join('');
  const count1 = [...map1.values()].sort().join('');
  const count2 = [...map2.values()].sort().join('');

  return letter1 === letter2 && count1 === count2;
};

var word1 = 'abc',
  word2 = 'bca';
var expected = true;
var result = closeStrings(word1, word2);
console.log(result, result === expected);

var word1 = 'a',
  word2 = 'aa';
var expected = false;
var result = closeStrings(word1, word2);
console.log(result, result === expected);

var word1 = 'cabbba',
  word2 = 'abbccc';
var expected = true;
var result = closeStrings(word1, word2);
console.log(result, result === expected);

File: 1601-1700/1651-1660/1657M_Determine_if_Two_Strings_Are_Close/cs/map.cs
// 1657. Determine if Two Strings Are Close
// https://leetcode.com/problems/determine-if-two-strings-are-close/description/
// T.C.: O(n)
// S.C.: O(1)
public class Solution
{
  public bool CloseStrings(string word1, string word2)
  {
    var word1Count = new Dictionary<char, int>();
    var word2Count = new Dictionary<char, int>();
    foreach (var c in word1)
    {
      if (word1Count.ContainsKey(c)) word1Count[c]++;
      else word1Count[c] = 1;
    }
    foreach (var c in word2)
    {
      if (word2Count.ContainsKey(c)) word2Count[c]++;
      else word2Count[c] = 1;
    }

    if (word1Count.Count != word2Count.Count) return false;

    var word1Freq = word1Count.Values.ToArray();
    var word2Freq = word2Count.Values.ToArray();
    Array.Sort(word1Freq);
    Array.Sort(word2Freq);
    return word1Freq.SequenceEqual(word2Freq);
  }
}

var word1 = "abc";
var word2 = "bca";
var expected = true;
var result = new Solution().CloseStrings(word1, word2);
Console.WriteLine($"{result}, {result == expected}");

word1 = "a";
word2 = "aa";
expected = false;
result = new Solution().CloseStrings(word1, word2);
Console.WriteLine($"{result}, {result == expected}");

word1 = "cabbba";
word2 = "abbccc";
expected = true;
result = new Solution().CloseStrings(word1, word2);
Console.WriteLine($"{result}, {result == expected}");
